### PR TITLE
Add ATX Header Support for `terraform-docs`

### LINKED ISSUE #196: Add a script to run `terraform-docs` and repair Markdown headers
# 💡 Summary #

We should add a script to this project that
1. Runs `terraform-docs .` from the root directory as well as any necessary submodule or example directories
2. Runs a `sed` command to repair the Markdown headers that `terraform-docs` leaves open

## Motivation and context ##

This would be useful because it would make updating the Terraform documentation as simple as running a single command.

### PR DESCRIPTION
**Pull request by michaelsaki**
## 🗣 Description ##

This pull request:

- Clones @mcdonnnj's forked branch from `terraform-docs`.
- Builds and installs the binary from the clone.

## 💭 Motivation and context ##

This temporarily resolves [#196](https://github.com/cisagov/skeleton-tf-module/issues/196) until @mcdonnnj's changes get merged into `terraform-docs`, at which point we can change back over to the standard of installing `terraform-docs`.

## 🧪 Testing ##

All automated tests passed. I also used [act](https://github.com/nektos/act) for local testing to make sure the workflow would behave like it should.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] If applicable, All future TODOs are captured in issues, which are referenced in the PR description.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

- [ ] Revert dependencies to default branches.
- [ ] Finalize version.

## ✅ Post-merge checklist ##

- [ ] Create a release.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -8,9 +8,12 @@ on:
     types: [apb]
 
 env:
+  BRANCH_NAME: improvement/support_atx_closed_markdown_headers
   CURL_CACHE_DIR: ~/.cache/curl
+  DEPTH: 1
   PIP_CACHE_DIR: ~/.cache/pip
   PRE_COMMIT_CACHE_DIR: ~/.cache/pre-commit
+  REPO_URL: https://github.com/mcdonnnj/terraform-docs.git
   RUN_TMATE: ${{ secrets.RUN_TMATE }}
```
Comment (mcdonnnj) [inline .github/workflows/build.yml]: I'd prefer these prepended to make it clear how they're being used.

```suggestion
  CURL_CACHE_DIR: ~/.cache/curl
  PIP_CACHE_DIR: ~/.cache/pip
  PRE_COMMIT_CACHE_DIR: ~/.cache/pre-commit
  RUN_TMATE: ${{ secrets.RUN_TMATE }}
  TERRAFORM_DOCS_REPO_BRANCH_NAME: improvement/support_atx_closed_markdown_headers
  TERRAFORM_DOCS_REPO_DEPTH: 1
  TERRAFORM_DOCS_REPO_URL: https://github.com/mcdonnnj/terraform-docs.git
```

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -124,11 +127,20 @@ jobs:
...
-          PACKAGE_VERSION: ${{ steps.setup-env.outputs.terraform-docs-version }}
-        run: go install ${PACKAGE_URL}@${PACKAGE_VERSION}
+      # We are temporarily using @mcdonnnj's forked branch of terraform-docs
+      # until his PR: https://github.com/terraform-docs/terraform-docs/pull/745
+      # is approved. This temporary fix will allow for ATX Header Support when
+      # terraform-docs is ran during lint.
+      - name: Clone ATX Headers branch from Terraform-docs Fork
+        run: |
+          git clone --branch $BRANCH_NAME --single-branch \
+           --depth=$DEPTH $REPO_URL /tmp/terraform-docs
```
Comment (mcdonnnj) [inline .github/workflows/build.yml]: In line with the previous suggestion.

```suggestion
          git clone --single-branch \
           --branch $TERRAFORM_DOCS_REPO_BRANCH_NAME \
           --depth $TERRAFORM_DOCS_REPO_DEPTH \
           $TERRAFORM_DOCS_REPO_URL /tmp/terraform-docs
```

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -124,11 +127,20 @@ jobs:
...
+      - name: Build and Install Terraform-docs Binary
+        run: |
+          cd /tmp/terraform-docs
+          GOBIN=$(go env GOPATH)/bin
+          go build -o $GOBIN/terraform-docs
+          echo "$GOBIN" >> $GITHUB_PATH
```
Comment (mcdonnnj) [inline .github/workflows/build.yml]: I believe the PATH is handled by [actions/setup-go](https://github.com/actions/setup-go/blob/0c52d547c9bc32b1aa3301fd7a9cb496313a4491/src/main.ts#L44). We can also take advantage of the `-C` switch to handle building from the cloned repository.

```suggestion
          go build \
          -C /tmp/terraform-docs \
          -o $(go env GOPATH)/bin/terraform-docs
```

**Review by mcdonnnj — CHANGES_REQUESTED** [review]
Thanks for working on this! Some feedback for your consideration.

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -124,11 +127,20 @@ jobs:
...
         run: go install ${PACKAGE_URL}@${PACKAGE_VERSION}
-      - name: Install Terraform-docs
-        env:
-          PACKAGE_URL: github.com/terraform-docs/terraform-docs
-          PACKAGE_VERSION: ${{ steps.setup-env.outputs.terraform-docs-version }}
-        run: go install ${PACKAGE_URL}@${PACKAGE_VERSION}
+      # We are temporarily using @mcdonnnj's forked branch of terraform-docs
+      # until his PR: https://github.com/terraform-docs/terraform-docs/pull/745
+      # is approved. This temporary fix will allow for ATX Header Support when
+      # terraform-docs is ran during lint.
```
Comment (jsf9k) [inline .github/workflows/build.yml]: Grammar:
```suggestion
      # terraform-docs is run during linting.
```

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -124,11 +127,20 @@ jobs:
...
           PACKAGE_VERSION: ${{ steps.setup-env.outputs.staticcheck-version }}
         run: go install ${PACKAGE_URL}@${PACKAGE_VERSION}
-      - name: Install Terraform-docs
-        env:
-          PACKAGE_URL: github.com/terraform-docs/terraform-docs
-          PACKAGE_VERSION: ${{ steps.setup-env.outputs.terraform-docs-version }}
-        run: go install ${PACKAGE_URL}@${PACKAGE_VERSION}
+      # We are temporarily using @mcdonnnj's forked branch of terraform-docs
+      # until his PR: https://github.com/terraform-docs/terraform-docs/pull/745
+      # is approved. This temporary fix will allow for ATX Header Support when
```
Comment (jsf9k) [inline .github/workflows/build.yml]: Remove unnecessary caps:
```suggestion
      # is approved. This temporary fix will allow for ATX header support when
```

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -124,11 +127,20 @@ jobs:
...
+      # terraform-docs is ran during lint.
+      - name: Clone ATX Headers branch from Terraform-docs Fork
```
Comment (jsf9k) [inline .github/workflows/build.yml]: Remove needless capitalization:
```suggestion
      - name: Clone ATX headers branch from terraform-docs fork
```

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -124,11 +127,20 @@ jobs:
...
+        run: |
+          git clone --branch $BRANCH_NAME --single-branch \
+           --depth=$DEPTH $REPO_URL /tmp/terraform-docs
+      - name: Build and Install Terraform-docs Binary
```
Comment (jsf9k) [inline .github/workflows/build.yml]: Remove needless capitals:
```suggestion
      - name: Build and install terraform-docs binary
```

**Review by jsf9k — CHANGES_REQUESTED** [review]
Some very minor things...

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -124,11 +127,21 @@ jobs:
...
         run: go install ${PACKAGE_URL}@${PACKAGE_VERSION}
-      - name: Install Terraform-docs
-        env:
-          PACKAGE_URL: github.com/terraform-docs/terraform-docs
-          PACKAGE_VERSION: ${{ steps.setup-env.outputs.terraform-docs-version }}
-        run: go install ${PACKAGE_URL}@${PACKAGE_VERSION}
+      # We are temporarily using @mcdonnnj's forked branch of terraform-docs
+      # until his PR: https://github.com/terraform-docs/terraform-docs/pull/745
+      # is approved. This temporary fix will allow for ATX header support when
+      # terraform-docs is run during linting.
```
Comment (dav3r) [inline .github/workflows/build.yml:134]: Please create an issue in this repo about changing from @mcdonnnj's fork back to the main `terraform-docs` repo after https://github.com/terraform-docs/terraform-docs/pull/745 is (hopefully) merged.  Once the issue has been created, please update the code comment here to be a `TODO` and include a link to that issue.  See [here](https://github.com/cisagov/cool-assessment-provisioner-iam/blob/3235e1f2db56423c52bb13f8460436ffb9d74de4/locals.tf#L46-L50) for an example.

File: .github/workflows/build.yml [hunk 9]
```diff
@@ -124,11 +127,21 @@ jobs:
...
+      - name: Clone ATX headers branch from terraform-docs fork
+        run: |
+          git clone --single-branch \
+           --branch $TERRAFORM_DOCS_REPO_BRANCH_NAME \
+           --depth $TERRAFORM_DOCS_REPO_DEPTH \
+           $TERRAFORM_DOCS_REPO_URL /tmp/terraform-docs
```
Comment (dav3r) [inline .github/workflows/build.yml]: I'm nitpicking here, but in our spirit of alphabetizing, might as well go all in and keep the switches in order:
```suggestion
          git clone \
           --branch $TERRAFORM_DOCS_REPO_BRANCH_NAME \
           --depth $TERRAFORM_DOCS_REPO_DEPTH \
           --single-branch \
           $TERRAFORM_DOCS_REPO_URL /tmp/terraform-docs
```

**Review by jsf9k — APPROVED** [review]

File: .github/workflows/build.yml (hunk 8 above)
Comment (michaelsaki) [reply .github/workflows/build.yml:134]: @dav3r Please see the changes here: https://github.com/cisagov/skeleton-generic/pull/164/commits/d114fb4233f22cb9edb3e1a6555ca0a677bcfc7e and also the issue created here: [165](https://github.com/cisagov/skeleton-generic/issues/165)

Comment (dav3r) [reply .github/workflows/build.yml:134]: The comment and issue both look great, thanks!  FYI, I made a couple of small formatting edits to #165.

**Review by dav3r — APPROVED** [review]
Thanks for taking care of this! 👍 

Can you please confirm that you re-tested this workflow after making the changes requested during the PR review?

**Comment by michaelsaki** [comment]
> Thanks for taking care of this! 👍
> 
> Can you please confirm that you re-tested this workflow after making the changes requested during the PR review?

You are welcome! Yes, I ran any changes to the workflow in [act](https://github.com/nektos/act), everything checked out locally.

**Review by mcdonnnj — APPROVED** [review]
:shipit:

**Pull request merged**